neural_network/2_hidden_layers_neural_network.py

```diff
@@ -0,0 +1,203 @@
+        updates the weights between the layers
+
+        >>> input_val =numpy.array(([0,0,0],[0,0,0],[0,0,0]),dtype=float)
+        >>> output_val = numpy.array(([0],[0],[0]),dtype=float)
+        >>> nn = NeuralNetwork(input_val,output_val)
+        >>> res = nn.feedforward()
+        >>> weights = nn.back_propagation()
+        """
+
+        updated_weights3 = numpy.dot(
+            self.layer2.T, 2 * (self.y - self.output) * sigmoid_derivative(self.output)
+        )
+        updated_weights2 = numpy.dot(
+            self.layer1.T,
+            numpy.dot(
+                2 * (self.y - self.output) * sigmoid_derivative(self.output),
+                self.weights3.T,
+            )
+            * sigmoid_derivative(self.layer2),
+        )
+        updated_weights1 = numpy.dot(
+            self.input.T,
+            numpy.dot(
+                numpy.dot(
+                    2 * (self.y - self.output) * sigmoid_derivative(self.output),
+                    self.weights3.T,
+                )
+                * sigmoid_derivative(self.layer2),
+                self.weights2.T,
+            )
+            * sigmoid_derivative(self.layer1),
+        )
+
+        self.weights1 += updated_weights1
+        self.weights2 += updated_weights2
+        self.weights3 += updated_weights3
+
+    def train(self, output: numpy.array, iterations: int, give_loss: bool) -> None:
+        """
+        output : required for calculating loss
+        performs the feeding and back propagation process for the
+        given number of iterations
+        every iteration will update the weights of neural network
+
+        >>> input_val = numpy.array(([0,0,0],[0,1,0],[0,0,1]),dtype=float)
+        >>> output_val = numpy.array(([0],[1],[1]),dtype=float)
+        >>> nn = NeuralNetwork(input_val,output_val)
+        >>> nn.train(output_val,10,False)
+        """
+        for iteration in range(1, iterations + 1):
+            self.output = self.feedforward()
+            self.back_propagation()
+            if give_loss:
+                print(
+                    "Iteration %s " % iteration,
+                    "Loss: "
+                    + str(numpy.mean(numpy.square(output - self.feedforward()))),
+                )
+
+    def predict(self, input: list) -> int:
+        """
+        predict's output for the given input values
+
+        >>> input_val =numpy.array(([0,0,0],[0,0,0],[0,0,0]),dtype=float)
+        >>> output_val =numpy.array(([0],[0],[0]),dtype=float)
+        >>> nn = NeuralNetwork(input_val,output_val)
+        >>> nn.train(output_val,100,False)
+        >>> nn.predict([0,0,0])
+        0
+        """
+        self.array = input
+        self.layer1 = sigmoid(numpy.dot(self.array, self.weights1))
+        self.layer2 = sigmoid(numpy.dot(self.layer1, self.weights2))
+        self.layer3 = sigmoid(numpy.dot(self.layer2, self.weights3))
+        if self.layer3 > 0.45:
+            return 1
+        else:
+            return 0
```

Comment:
What does the output values represent? Please mention this in the function contract (function docstring).

1 and 0 represent Truthy and Falsy values, so if you're going for that then you should probably return a boolean value.

Reply:
What I mean by the above comment is to mention what does 0 and 1 represent. If the user uses the function and gets 0 and 1, how does he/she know what to do with that?

Also, can you tell me why should we not return the value itself which is between 0 and 1 instead?

Reply:
The output is given in binary classification, '0' and '1' are the two classes provided in this training data.Our model provides a value in-between 0-1 so i have set a threshold value which assigns the value to one of the classes